### PR TITLE
additional test units, docs, bump is-glob

### PR DESCRIPTION
**Pull request by jonschlinkert**
Trying to make sure we have good coverage here, to future-proof against downstream changes as much as possible.

### CODE REVIEW

**Comment by es128** [comment]
Looks like I made the CI integrations very confused when I pushed a branch named `pr/7`

I've deleted it so that when this branch gets updated everything should go back to normal.

**Comment by jonschlinkert** [comment]
> I've deleted it so that when this branch gets updated everything should go back to normal.

do you know if that test you linked to was failing on the latest code? Or was it outdated?

**Comment by es128** [comment]
I checked out this PR, rebased it to master (to get the appveyor config), and pushed as a local branch. So it should have been up-to-date in terms of the dep tree when it ran those tests.

**Comment by jonschlinkert** [comment]
Ok, my hunch was that on windows, `path.dirname` is recognizing the `/\\` in `path/\\*(a|b)` as a single path separator, and it looks like that's exactly what's happening. We're looking at options now.

**Comment by phated** [comment]
I was wondering if this module doesn't need a custom `path.dirname` that only splits on `/` since that's the only valid glob separator

**Comment by es128** [comment]
That isn't quite my interpretation. I believe it is recognizing the `\\` as the separator, so the `/` is part of the dirname.

**Comment by es128** [comment]
> since that's the only valid glob separator

I'm concerned with other `path` operations being applied (like `path.join()`), leading to a mix of separators on windows even when the user input is valid.

Until support for glob escape characters entered the fray, there was no problem. Trying to support both may be hairy, and if it comes down to a choice I'm not sure that support for the escape chars is more important than the separator flexibility.

**Comment by es128** [comment]
Maybe the resolution for this and #8 simultaneously is:

```diff
diff --git a/index.js b/index.js
index 61615f1..de7f829 100644
--- a/index.js
+++ b/index.js
@@ -6,5 +6,5 @@ var isglob = require('is-glob');
 module.exports = function globParent(str) {
        str += 'a'; // preserves full path in case of trailing path separator
        do {str = path.dirname(str)} while (isglob(str));
-       return str;
+       return path.join(str, ' ').trim();
 };
```

```
> path.win32.join('path/', ' ').trim()
'path\\'
```

**Comment by phated** [comment]
@es128 I'm actually relying on not ending with a separator in my latest changes (so #8 won't need a fix).  We've been making a lot of changes to globbing packages because `\\` isn't a valid path character as per https://github.com/isaacs/node-glob#windows

**Comment by doowb** [comment]
Since globs should only have posix slashes (as @phated just pointed out in that link), we can use `path.posix.dirname(str)`. I've tested this locally on a windows machine and all of the tests pass. 

I think that it's acceptable to specify that this library will support escape characters in glob patterns, but not windows separators.

**Comment by phated** [comment]
@doowb that's a good solution but I don't think node 0.10 has `.posix` namespace exposed.

**Comment by jonschlinkert** [comment]
> that's a good solution but I don't think node 0.10 has .posix namespace exposed.

lol foiled again. damn you windows

**Comment by phated** [comment]
https://www.npmjs.com/package/path-posix ?

**Comment by doowb** [comment]
@phated i'm testing out `path-posix` now.

**Comment by jonschlinkert** [comment]
> I'm not sure that support for the escape chars is more important than the separator flexibility.

IMHO escape chars are more important that path separator flexibility, because by changing escapes we would be changing the actual intent of the glob pattern. Not just incorrectly joining. just my 2c

**Comment by jonschlinkert** [comment]
If necessary we can do a pr to path-posix, it hasn't been updated in 2 years, but many changes have been made to path handling in node.

**Comment by jonschlinkert** [comment]
> That isn't quite my interpretation. I believe it is recognizing the \\ as the separator, so the / is part of the dirname.

yes, you were correct there

**Comment by es128** [comment]
If we are going that way I think we should make a new one that is structured as a ponyfill. Potentially isolate it to just the `dirname` method.

**Comment by jonschlinkert** [comment]
> If we are going that way I think we should make a new one that is structured as a ponyfill. Potentially isolate it to just the dirname method.

that's very close to what @doowb said a little while ago. IMHO it seems like a last resort kind of option, but if we go that route would it make sense to focus the lib on globs? Do we need another regular filepath dirname lib?

**Comment by doowb** [comment]
The tests pass using `path-posix` on windows so there are options:

- use `path-posix` directly (submitting PRs if necessary)
- create ponyfill library for posix dirname

**Comment by doowb** [comment]
Was just commenting at the same time @jonschlinkert commented...

> would it make sense to focus the lib on globs

I think it would be a good idea to have a library that works like `path` but knows about globs and can handle doing things like `join`, `resolve`, and `dirname`.

**Comment by jonschlinkert** [comment]
Maybe we should just this one step at a time and just get the `dirname` issue resolved first. It might be necessary to do what @doowb is suggesting (and it's a great idea) but I think that lib would lend itself well to being incrementally implemented. just my 2c

**Comment by es128** [comment]
Why should it be a last resort? The whole point of it being a ponyfill is that it uses the native implementation when available and polyfills it if not. It doesn't seem to exist, and we have a use case. 

As for a whole suite of glob-aware `path` utilities - what would it do that's different from `path.posix`?(along with accompanying ponyfill(s) for support of node versions that don't expose it on an as-needed basis)

**Comment by jonschlinkert** [comment]
> Why should it be a last resort? The whole point of it being a ponyfill is that it uses the native implementation when available and polyfills it if not.

I was leaning towards @doowb's description, which wouldn't really be a ponyfill since it would be focused on glob paths.

**Comment by phated** [comment]
I couldn't actually find a `path-dirname` module and I think it should exist.  I'm really starting to treat node's stdlib as an antipattern in my modules.

**Comment by phated** [comment]
I don't see the difference between a path and a glob for many operations we are performing (I was actually thinking about this in relation to `to-absolute-glob` this morning).

**Comment by doowb** [comment]
> what would it do that's different from path.posix?

> I don't see the difference between a path and a glob for many operations we are performing

[This issue](https://github.com/es128/path-dirname/issues/1) shows some patterns that require glob aware methods.

But as @jonschlinkert  mentioned above, I think we can get something working for `dirname` then move on from there, if necessary.

**Comment by es128** [comment]
a more glob-aware version of path.posix.dirname ~= glob-parent

right?

By all means we should validate it on every weird edge case we can find, but if the ponyfill does the trick, then I think we're all set here. Remember we only needed to solve for backslash confusion from escape chars.

File: test.js [hunk 1]
```diff
@@ -5,57 +5,107 @@ var assert = require('assert');
...
     assert.equal(gp('path/**/*'), 'path');
     assert.equal(gp('path/**/subdir/foo.*'), 'path');
+    assert.equal(gp('path/subdir/**/foo.js'), 'path/subdir');
+    assert.equal(gp('path/!subdir/foo.js'), 'path/!subdir');
+  });
+
+  it('should respect escaped characters', function() {
+    assert.equal(gp('path/\\*\\*/subdir/foo.*'), 'path/\\*\\*/subdir');
+    assert.equal(gp('path/\\[\\*\\]/subdir/foo.*'), 'path/\\[\\*\\]/subdir');
+    assert.equal(gp('path/\\*(a|b)/subdir/foo.*'), 'path');
```
Comment (es128) [inline test.js]: this one can be disregarded now, but I'm interested in perspectives on the other questions.

File: README.md [hunk 2]
```diff
@@ -31,7 +33,22 @@ globParent('path/**/*'); // 'path'
...
+- `?` (question mark)
+- `*` (star)
+- `[` (opening bracket)
+- `]` (closing bracket)
+
+**Example**
+
+```js
+globParent('foo/[bar]/') // 'foo'
+globParent('foo/\\[bar]/') // 'foo/\\[bar]'
```
Comment (doowb) [inline README.md]: Jon and I were just talking about a method that would replace escaped glob characters with a heuristic, then run dirname, then restore the characters again. We can implement this in `path-dirname` and allow an option to be passed to run the logic for handling globs, but otherwise it can act as a polyfill to make it useful to other libraries.

**Comment by es128** [comment]
Here are CI results from this PR plus using the path-dirname ponyfill:

https://travis-ci.org/es128/glob-parent/builds/168658498
https://ci.appveyor.com/project/es128/glob-parent/build/1.0.8

You could apply commit ff51c15 to your fork in order to update this PR in kind and fix the failing test.

File: README.md (hunk 2 above)
Comment (jonschlinkert) [inline README.md]: why would we ever remove backslashes from escaped characters? That's what the problem is. Backslashes need to be retained so the globbing tool knows that characters are escaped.

Comment (es128) [inline README.md]: > why would we ever remove backslashes from escaped characters?

Because the output of glob-parent should be an actual path, free from any form of glob-ness and ready to pass into things like `fs.readdir`

Comment (jonschlinkert) [inline README.md]: oh I see, of course. Just the escaped characters in the dirname. I thought you meant the whole pattern for some reason (which obviously makes no sense in retrospect)

Comment (doowb) [inline README.md]: I think removing the escape characters in the final dirname makes sense here.

File: README.md [hunk 3]
```diff
@@ -31,7 +33,22 @@ globParent('path/**/*'); // 'path'
...
+```
+
+## Escaping
+
+The following characters have special significance in glob patterns and must be escaped if you want them to be treated as regular path characters:
 
+- `?` (question mark)
+- `*` (star)
+- `[` (opening bracket)
+- `]` (closing bracket)
```
Comment (es128) [inline README.md]: I also added pipe `|` in the escape-removal regex

anything else I might have missed?

Comment (es128) [inline README.md]: If you can update this PR with those other commits from my continuation branch and add the other escape-able chars to this list then I think we'll be ready to wrap this up.

Comment (jonschlinkert) [inline README.md]: great, I think @doowb is working on this now

**Comment by doowb** [comment]
@es128 I updated Jon's fork with your changes and added the additional characters to the readme.

It looks like github removed some of the inline code comments.

File: README.md [hunk 4]
```diff
@@ -41,6 +41,11 @@ The following characters have special significance in glob patterns and must be
 
 - `?` (question mark)
 - `*` (star)
+- `|` (pipe)
+- `(` (opening parenthese)
```
Comment (doowb) [inline README.md]: fixed, thanks
